### DISCUSSION TITLE
Window Freeze when click or moving the window on Windows platforms

### DISCUSSION THREAD

**Discussion by MartinLutherSu** [axmolengine/axmol#1343] [Help Desk]
On Windows systems, when the mouse is clicked on title bar of game window, or the game window is moved, the window contents will freeze. I know that this engine uses the Glew library to manage windows, and I know that this problem is a common problem of Glew on Windows systems. However, according to my idea, if multi-threading is used to separate the engine's main loop and Glew's monitoring mechanism, maybe this can be done? I made some attempts but failed. I was wondering if someone could consider completing this function. After all, similar engines such as Godot, MonoGame, Unity, and Unreal Engine have done this.

I am currently developing a game for the Windows platform using this engine. If the game content is not frozen when moving the window, it will bring a better experience to the user.

**Comment by halx99** [comment]
Not engine itself issue, but axmol support create render window as sub thread child window which x-studio UI editor does, it's create axmol window as child of mfc mainframe window and rendering in a unique thread

https://github.com/simdsoft/x-studio

**Comment by aismann** [comment]
> Not engine itself issue, but axmol support create render window as sub thread child window which x-studio UI editor does, it's create axmol window as mfc mainframe and rendering in a unique thread
> 
> https://github.com/simdsoft/x-studio

Maybe better a Q&A thema?

**Comment by MartinLutherSu** [comment]
Well, after my series of research, this seems to be the problem of glfw. If I want to solve this problem, it seems that I can only try to put the engine into the native framework of different platforms for rendering. For example, in the MFC framework of Windows 10, perhaps not using glfw. I am not sure this will work or not, and this seems to lose cross-platform functionality.

This got me thinking, and I was curious whether x-studio did the same thing? Unfortunately, this IDE is not open source.

**Comment by halx99** [comment]
This should be a long story, I'm not sure if it will support in the future, so I convert to discuss at this time

**Reply by MartinLutherSu** [reply]
Yeah alright, I have done it yesterday, I put the Axmol Runtime into a MFC framework, it works with some issues by now, it still need some fixes maybe. And I really have no idea about Linux and Mac, my game can not launch in these platforms, whatever, my game is not target for these platforms.... Perhaps someone noticed this topic can provide some good idea in the future...

**Reply by halx99** [reply]
psuedo code use exists glfw in axmol, Note: *The glfw only support create axmol game view as child window on win32 current, and it's not supported official glfw repo yet*
```cpp
glfwInit();
glfwWindowHint(GLFW_RESIZABLE, GL_TRUE);
//glfwWindowHint(GLFW_NO_WINDOW_CONTEXT, GLFW_TRUE);
glfwWindowHint(GLFW_CLIENT_API, GLFW_NO_API);

auto mainWindow = glfwCreateWindow(1280, 720, "axmol2-hwnd", nullptr, nullptr);


// create thread

std::thread renderThread([] {
    // setup appDelegate
    // GLContextAttrs has property 'viewParent' on win32 current
    // The normal axmol render loop at here, axmol game run in sub-thread
});

while (!glfwWindowShouldClose(mainWindow))
{
    glfwWaitEvents();
    printf("got gui event!\n");
}

if (renderThread.joinable())
    renderThread.join();
glfwTerminate();
```
